### DISCUSSION TITLE
Required extension VK_KHR_portability_enumeration is not available

### DISCUSSION THREAD

**Discussion by FlightBlaze** [DiligentGraphics/DiligentEngine#204] [Q&A] [answered]
I recently updated Diligent Engine to the latest version and came into a trouble with a Vulkan extension.

**Error:**

`Diligent Engine: ERROR in VulkanInstance() (VulkanInstance.cpp, 280): Required extension VK_KHR_portability_enumeration is not available`

**What I have done:**

`cd DiligentCore && git pull origin master && git submodule update --init --recursive`

Then I ran my project and got an error. After that, I updated Vulkan SDK to the latest version and after running my project got this error again.

**What I am doing (if you are curious):**

I am making software for automatic keyframe inbetweening for 2D animation. Basically, it takes two key frames of animation, some manually made guides and returns a lot of frames between these two key frames, creating the effect of motion.

**Comment by TheMostDiligent** [comment, marked as answer]
Is this on MacOS?

**Reply by TheMostDiligent** [reply]
> I updated Vulkan SDK to the latest version and after running my project got this error again

You need 1.3.216.0 SDK.
If you installed SDKs side-by-side, make sure that Diligent is using the most recent one. In particular, make sure that environment variables point to the right location. You may need to delete CMake cache and regenerate the project.
Look for CMake output, it should find the SDK and print the path:

```CMake
Using Vulkan library path: /Users/runner/VulkanSDK/macOS/lib
```

**Reply by FlightBlaze** [reply]
Yes, I am on macOS. I have 1.3.216.0 Vulkan SDK. All 5 environment variables point to the latest version of SDK. I also deleted build directory and generated Xcode project again. 

A fragment of CMake generation log:
`Using Vulkan library path: /Users/blaze/VulkanSDK/1.3.216.0/macOS/lib`

**Reply by TheMostDiligent** [reply]
Honestly, I don't know what the problem is. I just double-checked and everything works fine on both M1 and Intel GPUs.

Can you post full Diligent output - it should print the list of available layers and extensions:

```
Diligent Engine: Info: Available Vulkan instance layers: 
    VK_LAYER_LUNARG_api_dump 1.3.216
    VK_LAYER_KHRONOS_profiles 1.3.216
    VK_LAYER_KHRONOS_validation 1.3.216
    VK_LAYER_KHRONOS_synchronization2 1.3.216

Diligent Engine: Info: Supported Vulkan instance extensions: 
    VK_KHR_device_group_creation 0.0.1
    VK_KHR_external_fence_capabilities 0.0.1
    VK_KHR_external_memory_capabilities 0.0.1
    VK_KHR_external_semaphore_capabilities 0.0.1
    VK_KHR_get_physical_device_properties2 0.0.2
    VK_KHR_get_surface_capabilities2 0.0.1
    VK_KHR_surface 0.0.25
    VK_EXT_debug_report 0.0.10
    VK_EXT_debug_utils 0.0.2
    VK_EXT_metal_surface 0.0.1
    VK_EXT_swapchain_colorspace 0.0.4
    VK_MVK_macos_surface 0.0.3
    VK_KHR_portability_enumeration 0.0.1
```

Can you remove old SDK to make sure it is not being used somehow?

**Reply by FlightBlaze** [reply]
Here is the full log:

```
2022-07-19 04:01:20.720701+0300 tween[93398:8222231] [default] error finding potential wrapper bundle for node <FSNode 0x100517f70> { isDir = ?, path = '/Users/blaze/tween/build/Debug' }: Error Domain=NSOSStatusErrorDomain Code=-10811 "kLSNotAnApplicationErr: Item needs to be an application, but is not" UserInfo={_LSLine=1579, _LSFunction=wrapperBundleNodeForWrappedNode}
2022-07-19 04:01:20.720841+0300 tween[93398:8222231] [default] error finding potential wrapper bundle for node <FSNode 0x100517f70> { isDir = ?, path = '/Users/blaze/tween/build/Debug' }: Error Domain=NSOSStatusErrorDomain Code=-10811 "kLSNotAnApplicationErr: Item needs to be an application, but is not" UserInfo={_LSLine=1579, _LSFunction=wrapperBundleNodeForWrappedNode}
2022-07-19 04:01:20.720893+0300 tween[93398:8222231] [default] error finding potential wrapper bundle for node <FSNode 0x100517f70> { isDir = ?, path = '/Users/blaze/tween/build/Debug' }: Error Domain=NSOSStatusErrorDomain Code=-10811 "kLSNotAnApplicationErr: Item needs to be an application, but is not" UserInfo={_LSLine=1579, _LSFunction=wrapperBundleNodeForWrappedNode}
2022-07-19 04:01:21.326578+0300 tween[93398:8222231] [default] error finding potential wrapper bundle for node <FSNode 0x100517f70> { isDir = ?, path = '/Users/blaze/tween/build/Debug' }: Error Domain=NSOSStatusErrorDomain Code=-10811 "kLSNotAnApplicationErr: Item needs to be an application, but is not" UserInfo={_LSLine=1579, _LSFunction=wrapperBundleNodeForWrappedNode}
2022-07-19 04:01:21.451227+0300 tween[93398:8222231] Metal GPU Frame Capture Enabled
2022-07-19 04:01:21.451589+0300 tween[93398:8222231] Metal API Validation Enabled
Diligent Engine: Info: User-defined allocator is not provided. Using default allocator.

Diligent Engine: Info: Available Vulkan instance layers: 
    VK_LAYER_LUNARG_api_dump 1.2.189
    VK_LAYER_KHRONOS_validation 1.2.189
    VK_LAYER_LUNARG_device_simulation 1.2.189
    VK_LAYER_KHRONOS_synchronization2 1.2.189

Diligent Engine: Info: Supported Vulkan instance extensions: 
    VK_KHR_device_group_creation 0.0.1
    VK_KHR_external_fence_capabilities 0.0.1
    VK_KHR_external_memory_capabilities 0.0.1
    VK_KHR_external_semaphore_capabilities 0.0.1
    VK_KHR_get_physical_device_properties2 0.0.2
    VK_KHR_get_surface_capabilities2 0.0.1
    VK_KHR_surface 0.0.25
    VK_EXT_debug_report 0.0.10
    VK_EXT_debug_utils 0.0.2
    VK_EXT_metal_surface 0.0.1
    VK_EXT_swapchain_colorspace 0.0.4
    VK_MVK_macos_surface 0.0.3

Diligent Engine: ERROR in VulkanInstance() (VulkanInstance.cpp, 280): Required extension VK_KHR_portability_enumeration is not available

Unable to initialize Diligent Engine in Vulkan mode. The API may not be available, or required features may not be supported by this GPU/driver/OS version.
Program ended with exit code: 255
```

I also deleted the old Vulkan SDK and tried again. Still no `VK_KHR_portability_enumeration` extension.

M1 chip. macOS Monterey

**Reply by TheMostDiligent** [reply]
OK, that explains something - it loads the 1.2.189 version, not the latest one.
Is your environment configured as [explained here](https://github.com/DiligentGraphics/DiligentEngine#configuring-vulkan-build-environment)?

**Reply by TheMostDiligent** [reply]
Note that path to Vulkan SDK is required to find Vulkan loader. What version it loads is defined by ICDs listed in `VK_ICD_FILENAMES` environment variable

**Reply by FlightBlaze** [reply]
But I have only 1.3.216.0 version of SDK. How it loads the older one? I will reboot my Mac and try again

**Reply by TheMostDiligent** [reply]
> How it loads the older one? I will reboot my Mac and try again

I don't know:

```
Diligent Engine: Info: Available Vulkan instance layers: 
    VK_LAYER_LUNARG_api_dump 1.2.189
    VK_LAYER_KHRONOS_validation 1.2.189
    VK_LAYER_LUNARG_device_simulation 1.2.189
    VK_LAYER_KHRONOS_synchronization2 1.2.189
```
All layers are 1.2.189. Compare it with the output I pasted.

**Reply by FlightBlaze** [reply]
Hurray! Now it works! I uninstalled 1.2.189 using setup wizard, reinstalled 1.3.216 and now it works! Very thanks!

**Comment by thomsontg** [comment]
Hey,
I have the same issue and vilkaninfo gives version 1.3.250

What can I do?

**Reply by TheMostDiligent** [reply]
Are you using the latest version of Diligent?

**Reply by thomsontg** [reply]
Unfortunately, I am using my own render engine and not diligent. I posted it here because I couldn't find this issue anywhere else.

**Reply by TheMostDiligent** [reply]
Take a look here:

https://github.com/DiligentGraphics/DiligentCore/blob/ffa4b0d932112965134f1ae227014c63f0c88335/Graphics/GraphicsEngineVulkan/src/VulkanUtilities/VulkanInstance.cpp#L260

**Reply by thomsontg** [reply]
I have added VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME to the extension list which gives me ERROR_EXTENSION_NOT_PRESENT at the device creation however Without adding this extension, I get ERROR_INCOMPATIBLE_DRIVER.

Sorry, If I am explaining it badly, I am just getting started with vulkan.

**Reply by TheMostDiligent** [reply]
You can download and build Diligent Engine it see how it handles it.
